pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -253,8 +302,209 @@ void Stencil::CollectStrideOneIndices() {
+    return IsStrideOne(idx, tensorsOrder[1]) &&
+           IsStrideOne(idx, tensorsOrder[2]);
+  case 1:
+    // N is not restricted for stride one
+    return true;
+  case 2:
+    // Test if K is stride one for A(0)
+    return IsStrideOne(idx, tensorsOrder[0]);
+  default:
+    assert(!"Wrong matrix_idx");
+  }
+  return false;
+}
+
+bool Stencil::IndexExists(mlir::BlockArgument idx, unsigned tensor_idx) {
+  return tensorIdxs[tensor_idx].find(idx) != tensorIdxs[tensor_idx].end();
+}
+
+// Confirm if idx exists in the right place
+bool Stencil::ValidateIndexExistance(mlir::BlockArgument idx,
+                                     unsigned matrix_idx) {
+  switch (matrix_idx) {
+  case 0:
+    // Test if M exists in B and C, does not exist in A
+    return !IndexExists(idx, tensorsOrder[0]) && //
+           IndexExists(idx, tensorsOrder[1]) &&  //
+           IndexExists(idx, tensorsOrder[2]);
+  case 1:
+    // Test if N exists in A and C, does not exist in B
+    return IndexExists(idx, tensorsOrder[0]) &&  //
+           !IndexExists(idx, tensorsOrder[1]) && //
+           IndexExists(idx, tensorsOrder[2]);
+  case 2:
+    // Test if K exists in A and B, does not exist in C
+    return IndexExists(idx, tensorsOrder[0]) && //
+           IndexExists(idx, tensorsOrder[1]) && //
+           !IndexExists(idx, tensorsOrder[2]);
+  default:
+    assert(!"Wrong matrix_idx.");
+  }
+  return false;
+}
+
+// Search for matrix index (0 for M, 1 for N, 2 for K)
+void Stencil::SearchIndex(unsigned matrix_idx) {
+  if (matrix_idx >= kNumIndex) {
+    // We have the index and then search the tiles for these index
+    SearchTiles(0);
+    return;
+  }
+  auto &idxs = (matrix_idx == kNumIndex - 1) ? allIdxs : outIdxs;
+  for (auto idx : idxs) {
+    if (ValidateStrideOne(idx, matrix_idx) &&
+        ValidateIndexExistance(idx, matrix_idx)) {
+      innerIdxs[matrix_idx] = idx;
+      SearchIndex(nextMatrixIdx[matrix_idx]);
+    }
+  }
+}
+
+void Stencil::SearchTensorsOrder() {
+  // A B C, Search M(0) first as M is most restricted index
+  tensorsOrder[0] = 0;
+  tensorsOrder[1] = 1;
+  tensorsOrder[2] = 2;
+  SearchIndex(0);
+  // B A C, Search M(0) first as M is most restricted index
+  tensorsOrder[0] = 1;
+  tensorsOrder[1] = 0;
+  SearchIndex(0);
+}
+
+void Stencil::SearchTiles(unsigned idx) {
+  if (idx >= kNumIndex) {
+    double performance = Evaluate();
+    if (performance < bestPerf) {
+      bestPerf = performance;
+      for (unsigned i = 0; i < kNumTensors; ++i) {
+        bestTensorsOrder[i] = tensorsOrder[i];
+      }
+      for (unsigned i = 0; i < kNumIndex; ++i) {
+        bestIdxs[i] = innerIdxs[i];
+        bestTiles[i] = tiles[i];
+      }
+    }
+    return;
+  }
+
+  unsigned range = idxRange(innerIdxs[idx]);
+  for (unsigned i = range; i > 0; --i) {
+    if (onlyEven[idx] && (range % i != 0)) {
+      continue;
+    }
+    tiles[idx] = i;
+    SearchTiles(idx + 1);
+  }
+}
+
+int64_t Stencil::idxRange(mlir::BlockArgument idx) {
```

Comment:
We should probably cache or precompute this range.  Right now, we are recomputing this every time in the near innermost loop, and getConstantRanges is quite complex (involving affine simplification, etc).  I'm generally against early optimizations before measuring perf to be sure we care, but given that everything else is precomputed (is stride 1, etc), this one really stands out.  I think it's likely a relic of the fact the the range was just a simple constant in the pre-Affine version.

Reply:
Caching in the Stencil object.